### ISSUE TITLE
SPV Remapper: requires automated tests

### ISSUE THREAD

**Issue by steve-lunarg** [KhronosGroup/glslang#457]
To log the issue: the SPV remapper's testing has been non-public. It needs tests that are part of whichever project it lives in, both for basic correctness, and efficacy of compression improvement across a collection of shaders with partial commonality and different ID mappings.

**Comment by dneto0** [comment]
I strongly recommend using the SPIRV-Tools assembler as the basis for making test cases.  I worked on that assembler specifically so it could be the basis of SPIRV testing and conformance.  Credit goes to @kbenzie and Codeplay for kicking off the project and taking it forward a long way.   He also had the idea that the assembler could be used to make syntacitcally *invalid* SPIR-V.

In particular, a great many of the SPIRV-Tools tests are based on a single test utility class (test fixture) for  easily managing the assembly of a module and getting its results: https://github.com/KhronosGroup/SPIRV-Tools/blob/master/test/TestFixture.h#L45
(That makes up for the fact that we never publicized a C++ API for the assembler.)

**Comment by steve-lunarg** [comment]
Nod - I was thinking the same thing around use of SPIRV-Tools as a basis for testing.

**Comment by johnkslang** [comment]
Also consider having the remapper live as part of SPIRV-Tools instead of part of glslang.

**Comment by dneto0** [comment]
The SPIRV-Tools assembler is not ideal (today) to test the ID remapping algorithm. That's because it always assigns the next-higher ID value to each new name it encounters.  To test something like ID range compression (by removing gaps in the numbering), I'd have to extend the assembly language syntax to allow forced assignment of IDs.
E.g.
```
  %bool:500 = OpTypeBool   ;  will force the ID of %bool to be 500 instead of the assembler default
```

**Comment by dneto0** [comment]
The assembler does have a "force the next words to be exactly as I say" mode, but it's pretty cumbersome to use.

**Comment by steve-lunarg** [comment]
I've been thinking more about the disassembler and validator.  I can start with SPIR-V generated from canned GLSL shaders for instance, remap them, and compare the spirv-dis --raw-id --no-color post-remap version to a golden one.

In the past the idea of ID mapping invariant comparison has come up, though it's probably somewhat more involved to create that.  Anyway, I'm feeling like the remapper needs some TLC, and I plan to spend time on it soon.  Also, I have a fix coming for the defect @dneto0 found in my other fix.

**Comment by steve-lunarg** [comment]
BTW, the remapping algorithm actually expands the ID ranges.  Per design request, the algorithm operates on just one file at a time, so it has no ability to look at multiple files at once when calculating the ID maps.  (It can be given multiple files, but currently only handles them in the trivial repetition sense).  It rather attempts to canonicalize sequences with a process analogous to a convolution, which expands the ranges within certain defined limits.

**Comment by steve-lunarg** [comment]
I pushed a WIP branch #473 to solicit feedback on this issue.  It's got the mechanics, but isn't yet significantly populated.

Some points for discussion:

- I've made the remapper test a separate executable rather than including them in glslangtests.  On one hand, this means a little bit of duplication e.g, in the CMakeLists.txt, but if the remapper is to eventually be extracted from the glslang project, it seems like the less entanglement, the better.  (There is still some).  Similar for new C++ code in the test fixture.

- The strategy used for testing is to start with shaders in arbitrary source (HLSL, GLSL, even a SPIR-V binary), compile the shader to a SPIR-V blob, remap it under some set of remapper options which can vary from test to test, disassemble the post-remapper blob, and compare the disassembly text to a golden version.  We can also have canned SPV blobs for things not representable through GLSL or HLSL, e.g, to verify the 64 bit literals case.

- Currently, this disassembly uses the same disassembler than the other SPV tests in glslang use, rather than the SPIRV-Tools disassembler.  I felt that any interaction with another project should not be conflated with adding remapper tests.  Anyway that involves its own set of design decisions.

I will populate the tests in parallel with any discussion here, and PR it when any contentions are resolved.
